File: day05/day5.rb
def generate_map (file)
	map = Array.new
	while (line = file.gets(chomp: true)) =~ /[0-9]/
		numbers = line.split(" ").map {|n| Integer(n)}
		destRangeStart = numbers[0]
		sourceRangeStart = numbers[1]
		rangeSize = numbers[2]
		range = {:sourceStart => sourceRangeStart, :destStart => destRangeStart, :size => rangeSize}
		map.push(range)
	end
	return map
end

def get_map_value(map, index)
	value = index
	map.each do |range|
		sourceStart = range[:sourceStart]
		if index >= sourceStart and index < (sourceStart + range[:size])
			diff = index - sourceStart
			value = range[:destStart] + diff
		end
	end
	return value
end

def get_min_location
	File.open("day5Input.txt") do |f|
		seedsToBePlanted = f.readline.chomp!.split(":")[1].split(" ").map {|n| Integer(n)}
		f.readline()	# Empty line
		f.readline()	# seed-to-soil map header
		seedToSoil = generate_map(f)
		f.readline()	# Empty line
		soilToFert = generate_map(f)
		f.readline()	# Empty line
		fertToWater = generate_map(f)
		f.readline()	# Empty line
		waterToLight = generate_map(f)
		f.readline()	# Empty line
		lightToTemp = generate_map(f)
		f.readline()	# Empty line
		tempToHumidity = generate_map(f)
		f.readline()	# Empty line
		humidityToLocation = generate_map(f)
		minLocation = -1
		seedsToBePlanted.each do |seed|
			soil = get_map_value(seedToSoil, seed)
			fert = get_map_value(soilToFert, soil)
			water = get_map_value(fertToWater, fert)
			light = get_map_value(waterToLight, water)
			temp = get_map_value(lightToTemp, light)
			humidity = get_map_value(tempToHumidity, temp)
			location = get_map_value(humidityToLocation, humidity)
			if minLocation > location or minLocation == -1
				minLocation = location
			end
		end
		return minLocation
	end
end

def get_seed_ranges(file)
	map = Array.new
	seedNumbers = file.readline.chomp!.split(":")[1].split(" ").map {|n| Integer(n)}
	i = 0
	while i+1 < seedNumbers.length
		range = {:start => seedNumbers[i], :size => seedNumbers[i+1]}
		map.push(range)
		i += 2
	end
	# p map
	return map
end

def check_range_overlap(seedRanges, mapRanges)
	allNewRanges = Array.new
	# seedRanges.each do |seedRange|
	while !seedRanges.empty?
		seedRange = seedRanges.shift
		newRanges = Array.new
		mapRanges.each do |mapRange|
			seedRangeStart = seedRange[:start]
			seedRangeEnd = seedRangeStart + seedRange[:size]
			mapRangeStart = mapRange[:sourceStart]
			mapRangeEnd = mapRange[:sourceStart] + mapRange[:size]
			## Case 1: seed range is subset of map source range, map seed range start to value from map destination range
			if seedRangeStart >= mapRangeStart and seedRangeEnd <= mapRangeEnd
				rangeStartDiff = seedRangeStart - mapRangeStart
				newRange = {:start => mapRange[:destStart] + rangeStartDiff, :size => seedRange[:size]}
				newRanges.push(newRange)
				seedRanges.delete(seedRange)
				break
			## Case 2/3: partial overlap of ranges, split into two ranges
			elsif seedRangeStart < mapRangeStart and mapRangeStart < seedRangeEnd and seedRangeEnd < mapRangeEnd
				overlapSize = seedRangeEnd - mapRangeStart
				overlapRange = {:start => mapRange[:destStart], :size => overlapSize}
				newRanges.push(overlapRange)
				remainderRangeSize = mapRangeStart - seedRangeStart
				remainderRange = {:start => seedRangeStart, :size => remainderRangeSize}
				seedRanges.push(remainderRange)
				seedRanges.delete(seedRange)
				break
			elsif mapRangeStart < seedRangeStart and seedRangeStart < mapRangeEnd and mapRangeEnd < seedRangeEnd
				overlapSize = mapRangeEnd - seedRangeStart
				rangeStartDiff = seedRangeStart - mapRangeStart
				overlapRange = {:start => mapRange[:destStart] + rangeStartDiff, :size => overlapSize}
				newRanges.push(overlapRange)
				remainderRangeSize = seedRangeEnd - mapRangeEnd
				remainderRange = {:start => mapRangeEnd, :size => remainderRangeSize}
				seedRanges.push(remainderRange)
				seedRanges.delete(seedRange)
				break
			## Case 4: map source range is subset of seed range, split into three ranges
			elsif mapRangeStart >= seedRangeStart and mapRangeEnd <= seedRangeEnd
				overlapRange = {:start => mapRange[:destStart], :size => mapRange[:size]}
				newRanges.push(overlapRange)
				beforeRemainderRangeSize = mapRangeStart - seedRangeStart
				unless beforeRemainderRangeSize == 0
					beforeRemainderRange = {:start => seedRangeStart, :size => beforeRemainderRangeSize}
					seedRanges.push(beforeRemainderRange)
				end
				afterRemainderRangeSize = seedRangeEnd - mapRangeEnd
				unless afterRemainderRangeSize == 0
					afterRemainderRange = {:start => mapRangeEnd, :size => afterRemainderRangeSize}
					seedRanges.push(afterRemainderRange)
				end
				seedRanges.delete(seedRange)
				break
			end
		end
		## If no overlaps found with seedRange, add as is
		if newRanges.empty?
			newRanges.push(seedRange)
		end
		allNewRanges.concat(newRanges)
	end
	return allNewRanges
end

def get_min_location_ranges
	File.open("day5Input.txt") do |f|
		seedRangesToBePlanted = get_seed_ranges(f)
		f.readline()	# Empty line
		f.readline()	# seed-to-soil map header
		seedToSoil = generate_map(f)
		f.readline()	# Empty line
		soilToFert = generate_map(f)
		f.readline()	# Empty line
		fertToWater = generate_map(f)
		f.readline()	# Empty line
		waterToLight = generate_map(f)
		f.readline()	# Empty line
		lightToTemp = generate_map(f)
		f.readline()	# Empty line
		tempToHumidity = generate_map(f)
		f.readline()	# Empty line
		humidityToLocation = generate_map(f)
		soilRanges = check_range_overlap(seedRangesToBePlanted, seedToSoil)
		fertRanges = check_range_overlap(soilRanges, soilToFert)
		waterRanges = check_range_overlap(fertRanges, fertToWater)
		lightRanges = check_range_overlap(waterRanges, waterToLight)
		tempRanges = check_range_overlap(lightRanges, lightToTemp)
		humidityRanges = check_range_overlap(tempRanges, tempToHumidity)
		locationRanges = check_range_overlap(humidityRanges, humidityToLocation)
		locationRanges.sort_by! {|range| range[:start]}
		return locationRanges[0][:start]
	end
end

p get_min_location
p get_min_location_ranges

File: day05/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
	"slices"
)

type conversionMap struct {
	sourceStart, destStart, size int
}

type valueRange struct {
	start, size int
}

func print_error(err error) {
	if err != nil {
		fmt.Println(err)
	}
}

func generate_map(scanner *bufio.Scanner) []conversionMap {
	mapArray := []conversionMap{}
	scanner.Scan()	// Header line
	for scanner.Scan() {
		line := scanner.Text()
		// Empty lines between maps
		if line == "" {
			break
		}
		numbers := strings.Split(line, " ")
		destRangeStart, err := strconv.Atoi(numbers[0])
		print_error(err)
		sourceRangeStart, err := strconv.Atoi(numbers[1])
		print_error(err)
		rangeSize, err := strconv.Atoi(numbers[2])
		print_error(err)
		convMap := conversionMap{sourceStart: sourceRangeStart, destStart: destRangeStart, size: rangeSize}
		mapArray = append(mapArray, convMap)
	}
	return mapArray
}

func get_map_value(conversionMaps []conversionMap, number int) int {
	value := number
	for _, convMap := range conversionMaps {
		sourceStart := convMap.sourceStart
		size := convMap.size
		if number >= sourceStart && number < (sourceStart + size) {
			diff := number - sourceStart
			value = convMap.destStart + diff
		}
	}
	return value
}

func get_min_location() {
	file, err := os.Open("day5Input.txt")
	print_error(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	// Seeds
	scanner.Scan()
	seedsLine := scanner.Text()
	seedStrs := strings.Split(strings.Split(seedsLine, ": ")[1], " ")
	seeds := make([]int, len(seedStrs))
	for i, str := range seedStrs {
		seeds[i], err = strconv.Atoi(str)
		print_error(err)
	}
	scanner.Scan()  // Empty line between seeds and seedToSoil map
	seedToSoil := generate_map(scanner)
	soilToFert := generate_map(scanner)
	fertToWater := generate_map(scanner)
	waterToLight := generate_map(scanner)
	lightToTemp := generate_map(scanner)
	tempToHumid := generate_map(scanner)
	humidToLoc := generate_map(scanner)
	minLocation := -1
	for _, seed := range seeds {
		soil := get_map_value(seedToSoil, seed)
		fert := get_map_value(soilToFert, soil)
		water := get_map_value(fertToWater, fert)
		light := get_map_value(waterToLight, water)
		temp := get_map_value(lightToTemp, light)
		humid := get_map_value(tempToHumid, temp)
		loc := get_map_value(humidToLoc, humid)
		if minLocation > loc || minLocation == -1 {
			minLocation = loc
		}
	}
	fmt.Println(minLocation)
}

func check_range_overlap(conversionMaps []conversionMap, valRanges []valueRange) []valueRange {
	allNewRanges := []valueRange{}
	for len(valRanges) != 0 {
		valRange := valRanges[0]
		valRanges = valRanges[1:]
		newRanges := []valueRange{}
		for _, convMap := range conversionMaps {
			valRangeStart := valRange.start
			valRangeEnd := valRangeStart + valRange.size
			mapRangeStart := convMap.sourceStart
			mapRangeEnd := mapRangeStart + convMap.size
			// Case 1: seed range is subset of map source range, map seed range start to value from map destination range
			if valRangeStart >= mapRangeStart && valRangeEnd <= mapRangeEnd {
				rangeStartDiff := valRangeStart - mapRangeStart
				newRange := valueRange{start: convMap.destStart + rangeStartDiff, size: valRange.size}
				newRanges = append(newRanges, newRange)
				break
			// Case 2/3: partial overlap of ranges, split into two ranges
			} else if valRangeStart < mapRangeStart && mapRangeStart < valRangeEnd && valRangeEnd < mapRangeEnd {
				overlapSize := valRangeEnd - mapRangeStart
				overlapRange := valueRange{start: convMap.destStart, size: overlapSize}
				newRanges = append(newRanges, overlapRange)
				remainderSize := mapRangeStart - valRangeStart
				remainderRange := valueRange{start: valRangeStart, size: remainderSize}
				valRanges = append(valRanges, remainderRange)
				break
			} else if mapRangeStart < valRangeStart && valRangeStart < mapRangeEnd && mapRangeEnd < valRangeEnd {
				overlapSize := mapRangeEnd - valRangeStart
				rangeStartDiff := valRangeStart - mapRangeStart
				overlapRange := valueRange{start: convMap.destStart + rangeStartDiff, size: overlapSize}
				newRanges = append(newRanges, overlapRange)
				remainderSize := valRangeEnd - mapRangeEnd
				remainderRange := valueRange{start: mapRangeEnd, size: remainderSize}
				valRanges = append(valRanges, remainderRange)
				break
			// Case 4: map source range is subset of seed range, split into three ranges
			} else if mapRangeStart >= valRangeStart && mapRangeEnd <= valRangeEnd {
				overlapRange := valueRange{start: convMap.destStart, size: convMap.size}
				newRanges = append(newRanges, overlapRange)
				beforeRemainderSize := mapRangeStart - valRangeStart
				if beforeRemainderSize != 0 {
					beforeRemainderRange := valueRange{start: valRangeStart, size: beforeRemainderSize}
					valRanges = append(valRanges, beforeRemainderRange)
				}
				afterRemainderSize := valRangeEnd - mapRangeEnd
				if afterRemainderSize != 0 {
					afterRemainderRange := valueRange{start: mapRangeEnd, size: afterRemainderSize}
					valRanges = append(valRanges, afterRemainderRange)
				}
				break
			}
		}
		if len(newRanges) == 0 {
			newRanges = append(newRanges, valRange)
		}
		allNewRanges = append(allNewRanges, newRanges...)
	}
	return allNewRanges
}

func get_min_location_ranges() {
	file, err := os.Open("day5Input.txt")
	print_error(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	// Seed ranges
	scanner.Scan()
	seedsLine := scanner.Text()
	seedStrs := strings.Split(strings.Split(seedsLine, ": ")[1], " ")
	seedNums := make([]int, len(seedStrs))
	seedRanges := []valueRange{}
	for i, str := range seedStrs {
		seedNums[i], err = strconv.Atoi(str)
		print_error(err)
	}
	i := 0
	for i+1 < len(seedNums) {
		seedRanges = append(seedRanges, valueRange{start: seedNums[i], size: seedNums[i+1]})
		i += 2
	}
	scanner.Scan()  // Empty line between seeds and seedToSoil map
	seedToSoil := generate_map(scanner)
	soilToFert := generate_map(scanner)
	fertToWater := generate_map(scanner)
	waterToLight := generate_map(scanner)
	lightToTemp := generate_map(scanner)
	tempToHumid := generate_map(scanner)
	humidToLoc := generate_map(scanner)
	cmpStart := func(r1 valueRange, r2 valueRange) int {
		if r1.start < r2.start {
			return -1
		} else if r1.start > r2.start {
			return 1
		} else {
			return 0
		}
	}
	soilRanges := check_range_overlap(seedToSoil, seedRanges)
	fertRanges := check_range_overlap(soilToFert, soilRanges)
	waterRanges := check_range_overlap(fertToWater, fertRanges)
	lightRanges := check_range_overlap(waterToLight, waterRanges)
	tempRanges := check_range_overlap(lightToTemp, lightRanges)
	humidRanges := check_range_overlap(tempToHumid, tempRanges)
	locRanges := check_range_overlap(humidToLoc, humidRanges)
	slices.SortFunc(locRanges, cmpStart)
	fmt.Println(locRanges[0].start)
}

func main() {
	get_min_location()
	get_min_location_ranges()
}